src/ws/cert-session.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This program opens a session for a given user and runs the bridge in
+ * it, using TLS client certificate authentication and mapping it to a user
+ * through sssd (https://www.freeipa.org/page/V4/User_Certificates).
+ */
+
+#include "config.h"
+#include <stdlib.h>
+
+#include "session-utils.h"
+
+static char *last_txt_msg = NULL;
+static char *conversation = NULL;
+
+
+static int
+pam_conv_func (int num_msg,
+               const struct pam_message **msg,
+               struct pam_response **ret_resp,
+               void *appdata_ptr)
+{
+  /* we don't expect (nor can handle) any actual auth conversation here, but
+   * PAM sometimes sends messages like "Creating home directory for USER" */
+  for (int i = 0; i < num_msg; ++i)
+      debug ("got PAM conversation message, ignoring: %s", msg[i]->msg);
+  return PAM_CONV_ERR;
+}
+
+static pam_handle_t *
+perform_tlscert (const char *rhost,
```

Comment:
Is there any reason that once this is complete, than we can't move this code into cockpit-session? It's fine for now as a separate package, but lets aim to get it right later.